content/struct-classes.md

```diff
@@ -0,0 +1,502 @@
+  def this(name: String, age: Int): User = this(name, age, None)
+  private def copy(name: String = name, age: Int = age, email: Option[String] = email): User = new User(name, age, email)
+  // structural implementation of "toString", "equals", and "hashCode"
+  override def toString(): String = s"User($name, $age, $email)"
+  override def equals(that: Any): Boolean =
+    that match
+      case user: User => user.name == name && user.age == age && user.email == email
+      case _ => false
+  override def hashCode(): Int =
+    37 * (37 * (37 * (17 + name.##) + age.##) + email.##)
+  // transformation methods
+  def withName(newName: String): User = copy(name = newName)
+  def withAge(newAge: Int): User = copy(age = newAge)
+  def withEmail(newEmail: Option[String]): User = copy(email = newEmail)
+~~~
+
+Adding one field requires adding it to the `private` constructor, to the `private`
+method `copy`, to the `toString` implementation, to the `equals` implementation,
+to the `hashCode` implementation, and to add a transformation method (here,
+`withEmail`). This arguably a lot of effort.
+
+Are there any other solutions?
+
+### Using Code Generation or Meta-Programming
+
+Other solutions used by the community are based on code generation or meta-programming.
+
+[Contraband](https://eed3si9n.com/contraband-an-alternative-to-case-class/) is an
+example of tool that generates "case class like" definitions from an _ad-hoc_
+interface definition language. Approaches based on code generation have the following
+drawbacks:
+- the definitions of the data types use another language than Scala that you need
+  to learn and get familiar with the surrounding tooling
+- they require customizing the build definition to correctly generate the sources
+- they are often not well supported by IDEs
+- they make the code harder to navigate through.
+
+The last type of solutions are based on meta-programming facilities.
+[data-class](https://github.com/alexarchambault/data-class) and
+[scalameta](https://github.com/scalameta/scalameta/blob/01cb1137cac89d1453846ef1e7acb8f4a8833e6c/scalameta/common/shared/src/main/scala/org/scalameta/data/data.scala)
+are two examples of such solutions. They seem to be hard to maintain(see e.g.
+[data-class#120](https://github.com/alexarchambault/data-class/issues/120) and
+[scalameta#2485](https://github.com/scalameta/scalameta/issues/2485)). Also, IDE
+support of meta-programming-based approaches is not always good.
+
+### Scope of the Proposal
+
+We would like to support, at the language level, the definition of data types
+that would support the following features:
+- can evolve in a binary compatible way (ie, developers can define a new version
+  of the data type, with added or removed fields, in a backward binary compatible
+  way)
+- structural implementation of `equals`, `hashCode`, `toString`
+- support of JVM-based serialization
+- primary constructor parameters promoted to public fields
```

Comment:
Thanks, Julien, for looking for a solution to the bin compact issues we face with case classes.
Would `struct` classes implement `Product` as case classes do?

Reply:
I don’t have a strong opinion about that. What would be the motivation for implementing `Product`?

Reply:
The motivation would be that struct classes work with derived Mirror instances, which include `fromProduct`.

Reply:
Mirrors of struct classes could create compatibility issues: consider code compiled with a mirror synthesized for a version x of a data type, and code compiled with a mirror synthesized for a version y of the same data type. Both program may refer to the “same” type from those mirrors, but those types may be different in practice, leading to runtime issues.